src/Microsoft.ML.Data/Prediction/Calibrator.cs

```diff
@@ -116,21 +116,44 @@ internal interface ISelfCalibratingPredictor
         IPredictor Calibrate(IChannel ch, IDataView data, ICalibratorTrainer caliTrainer, int maxRows);
     }
 
-    public abstract class CalibratedPredictorBase :
+    /// <summary>
+    /// <see cref="IHasWeeklyTypedCalibratedPredictor"/> provides a weekly-typed way to access strongly-typed
+    /// <see cref="CalibratedPredictorBase{TSubPredictor, TCalibrator}.SubPredictor"/> and
+    /// <see cref="CalibratedPredictorBase{TSubPredictor, TCalibrator}.Calibrator"/>.
+    /// <see cref="IHasWeeklyTypedCalibratedPredictor"/> is commonly used in weekly-typed expressions. The
+    /// existence of this interface is just for supporting existing codebase, so we discourage its uses.
+    /// </summary>
+    [BestFriend]
+    internal interface IHasWeeklyTypedCalibratedPredictor
```

Comment:
>IHasWeeklyTypedCalibratedPredictor [](start = 23, length = 34)

Clever.

Reply:
Though, I think probably you *could* have gotten away with just making a couple `[BestFriend] internal` fields on the calibrated predictor base.

---
In reply to: [254540575](https://github.com/dotnet/machinelearning/pull/2435#discussion_r254540575) [](ancestors = 254540575)

Reply:
I am not sure if we can do that. There are some places where we need to enumerate ALL possible combinations of `TSubPredictor` and `TCalibrator` in `CalibratedPredictorBase` to cast an object to the desired type; for example, in [SummarizePredictor](https://github.com/dotnet/machinelearning/blob/3dd3a1ebaff92dba34c2bb2ecd833c97beae1a23/src/Microsoft.ML.Data/EntryPoints/SummarizePredictor.cs#L51) we have
```csharp
        internal static IDataView GetSummaryAndStats(IHostEnvironment env, IPredictor predictor, RoleMappedSchema schema, out IDataView stats)
        {
            var calibrated = predictor as CalibratedPredictorBase;
            while (calibrated != null)
            {
                predictor = calibrated.SubPredictor;
                calibrated = predictor as CalibratedPredictorBase;
            }
            ...
       }
```
Removing `IHasWeeklyTypedCalibratedPredictor` may force us to extend that function significantly.

---
In reply to: [254541286](https://github.com/dotnet/machinelearning/pull/2435#discussion_r254541286) [](ancestors = 254541286,254540575)

Reply:
OK, that's fine, just throwing it out there. Thanks @wschin !